contracts/ABSBlockRelay.sol

```diff
@@ -0,0 +1,398 @@
+  uint256 activeIdentities;
+
+  ABSInterface wbi;
+
+  // Last block reported
+  Beacon public lastBlock;
+
+  // Map a vote proposed to the number of votes received and its hashes
+  mapping(uint256=> VoteInfo) internal voteInfo;
+
+  // Map the hash of the block to the merkle roots and the previousVote it extends
+  mapping (uint256 => MerkleRoots) public blocks;
+
+  // Map an epoch to the finalized blockHash
+  mapping(uint256 => uint256) internal epochFinalizedBlock;
+
+  constructor(
+    uint256 _witnetGenesis, uint256 _epochSeconds, uint256 _firstBlock, address _wbiAddress) public{
+    // Set the first epoch in Witnet plus the epoch duration when deploying the contract
+    witnetGenesis = _witnetGenesis;
+    epochSeconds = _epochSeconds;
+    firstBlock = _firstBlock;
+    wbi = ABSInterface(_wbiAddress);
+  }
+
+  // Ensure block exists
+  modifier blockExists(uint256 _id){
+    require(blocks[_id].drHashMerkleRoot!=0, "Non-existing block");
+    _;
+  }
+  // Ensure block does not exist
+  modifier blockDoesNotExist(uint256 _id){
+    require(blocks[_id].drHashMerkleRoot==0, "The block already existed");
+    _;
+  }
+
+  // Ensure that neither Poi nor PoE are allowed if the epoch is pending
+  modifier epochIsFinalized(uint256 _epoch){
+    require(
+      (epochFinalizedBlock[_epoch] != 0),
+      "The block has not been finalized");
+    _;
+  }
+
+  // Ensure that the msg.sender is in the abs
+  modifier isAbsMember(address _address){
+
+    require(wbi.isABSMember(_address) == true, "Not a member of the abs");
+    _;
+  }
+
+  // Ensure the epoch for which the block is been proposed is valid
+  // Valid if it is one epoch before the current epoch
+  modifier epochValid(uint256 _epoch){
+    currentEpoch = updateEpoch();
+    if (proposalEpoch == 0) {
+      proposalEpoch = currentEpoch;
+    }
+    require(currentEpoch - 1 == _epoch, "Proposing a block for a non valid epoch");
+    _;
+  }
+
+  /// @dev Retrieve the requests-only merkle root hash that was reported for a specific block header.
+  /// @param _blockHash Hash of the block header
+  /// @return Requests-only merkle root hash in the block header.
+  function readDrMerkleRoot(uint256 _blockHash)
+    external
+    view
+    blockExists(_blockHash)
+  returns(uint256 drMerkleRoot)
+    {
+    drMerkleRoot = blocks[_blockHash].drHashMerkleRoot;
+  }
+
+  /// @dev Retrieve the tallies-only merkle root hash that was reported for a specific block header.
+  /// @param _blockHash Hash of the block header
+  /// tallies-only merkle root hash in the block header.
+  function readTallyMerkleRoot(uint256 _blockHash)
+    external
+    view
+    blockExists(_blockHash)
+  returns(uint256 tallyMerkleRoot)
+  {
+    tallyMerkleRoot = blocks[_blockHash].tallyHashMerkleRoot;
+  }
+
+  /// @dev Verifies if the contract is upgradable
+  /// @return true if the contract upgradable
+  function isUpgradable() external pure returns(bool) {
```

Comment:
I'd reformulate this into something that actually performs the upgradeability validation.